src/wallet/wallet.cpp

```diff
@@ -6826,3 +6826,159 @@ void CWallet::PrecomputeSpends()
+
+
+void LinkAutoSpendThreadGroup(void* pthreadgroup)
+{
+    pthreadGroupAutoSpend = (boost::thread_group*)pthreadgroup;
+}
+
+bool StartAutoSpend()
+{
+    if (!pthreadGroupAutoSpend) {
+        error("%s: pthreadGroupAutoSpend is null! Cannot autospend.", __func__);
+        return false;
+    }
+
+    // Close any active precomputing threads before starting new threads
+    if (pthreadGroupAutoSpend->size() > 0) {
+        StopAutoSpend();
+    }
+
+    pthreadGroupAutoSpend->create_thread(boost::bind(&AutoSpendZeroCoin));
+    return true;
+}
+
+void StopAutoSpend()
+{
+    if (!pthreadGroupAutoSpend) {
+        error("%s: pthreadGroupAutoSpend is null! Cannot stop autospending.", __func__);
+        return;
+    }
+
+    if (pthreadGroupAutoSpend->size() > 0) {
+        pthreadGroupAutoSpend->interrupt_all();
+        pthreadGroupAutoSpend->join_all();
+    }
+
+    LogPrintf("AutoSpendZeroCoin stopped\n");
+}
+
+static libzerocoin::CoinDenomination denom_to_spend = libzerocoin::CoinDenomination::ZQ_TEN;
+static int max_number_to_spend = 1;
+
+void SetAutoSpendParameters(const int& nCount, const int& nDenom)
+{
+    if (nCount >= 1 && nCount <= 3)
+        max_number_to_spend = nCount;
+
+    libzerocoin::CoinDenomination denom = libzerocoin::IntToZerocoinDenomination(nDenom);
+
+    if (denom != libzerocoin::CoinDenomination::ZQ_ERROR)
+        denom_to_spend = denom;
+}
+
+void AutoSpendZeroCoin()
+{
+    LogPrintf("AutoSpendZeroCoin started: spending at most %d of the %d denomination per cycle\n", max_number_to_spend, libzerocoin::ZerocoinDenominationToInt(denom_to_spend));
+
+    static CTxDestination destination;
+
+    boost::this_thread::interruption_point();
+    try {
+        uint nMilliSeconds = 1000;
+        int count = 0;
+
+        while (true) {
+            // Sleep 5 minutes between spends, but actively check to see if the thread has been interrupted every 5 seconds
+            if (count < 300) {
+                count++;
+                boost::this_thread::interruption_point();
+                MilliSleep(nMilliSeconds);
+                boost::this_thread::interruption_point();
+                continue;
+            }
+            count = 0;
+
+            boost::this_thread::interruption_point();
+
+            if (ShutdownRequested())
+                break;
+
+            if (IsInitialBlockDownload() || !HeadersAndBlocksSynced()) {
+                LogPrintf("%s: waiting for sync cannot auto spend\n", __func__);
+                continue;
+            }
+
+            auto pwallet = GetMainWallet();
+
+            if (!pwallet) {
+                LogPrintf("%s: pwallet is null cannot auto spend\n", __func__);
+                continue;
+            }
+
+            if (pwallet->IsLocked() || pwallet->IsUnlockedForStakingOnly()) {
+                LogPrintf("%s: pwallet is locked cannot auto spend\n", __func__);
+                continue;
+            }
+
+            if (destination.empty()) {
```

Comment:
What is the reason for adding a key?

Reply:
When spending zerocoin, we want to make sure that we spend it to an address that the wallet owns. This only happens on the first loop of the auto spend, and uses the same address for as long as the auto spend thread runs.

Reply:
We should only spend to stealth addresses (not basecoin), and should only create an address to spend to once. Creating a stealth address and adding it to the address book should do the trick.

Reply:
Updated to spend to stealth addresses. Also, instead of trying to create a new address every time the wallet loads, the wallet creates and stores a stealth address to be used for auto spending. the user can change this address by using the **startautospend** rpc call